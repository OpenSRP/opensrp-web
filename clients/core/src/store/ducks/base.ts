import { get, keyBy, values, Dictionary } from 'lodash';
import { AnyAction, Store } from 'redux';
import SeamlessImmutable from 'seamless-immutable';

/** describes primary required properties for a client object from opensrp
 * extend this to create generic types for clients.
 */
export interface BaseClient {
    type: 'Client';
    baseEntityId: string;
}

// actions

/** CLIENTS_FETCHED action type */
export const CLIENTS_FETCHED = 'opensrp/reducer/clients/CLIENTS_FETCHED';
/** REMOVE_CLIENTS action type */
export const REMOVE_CLIENTS = 'opensrp/reducer/clients/REMOVE_CLIENTS';

/** interface for authorize action
 * ClientType - generic type - client type being handled by this function
 */
export interface FetchClientsAction<ClientType> extends AnyAction {
    clientsById: Dictionary<ClientType>;
    type: typeof CLIENTS_FETCHED;
}

/** Interface for removeClientsAction */
interface RemoveClientsAction extends AnyAction {
    clientsById: {};
    type: typeof REMOVE_CLIENTS;
}

/** Create type for clients reducer actions */
export type ClientsActionTypes<ClientType> = FetchClientsAction<ClientType> | RemoveClientsAction | AnyAction;

// factory methods for action creators

/** creates the action creator
 * ClientType - generic type - client type being handled by this function
 */
export function fetchClientsFactory<ClientType extends BaseClient>() {
    /** Fetch clients action creator
     * @param {Client []} clientsList - clients array to add to store
     * @return {FetchClientsAction} - an action to add clients to redux store
     */
    return (clientsList: ClientType[] = []): FetchClientsAction<ClientType> => ({
        clientsById: keyBy<ClientType>(clientsList, (client: ClientType) => client.baseEntityId),
        type: CLIENTS_FETCHED,
    });
}

// actions

/** removeClientsAction action */
export const removeClientsAction: RemoveClientsAction = {
    clientsById: {},
    type: REMOVE_CLIENTS,
};

// The reducer

/** interface for clients state in redux store
 * ClientType - generic type - client type being handled by this function
 */
interface ClientState<ClientType> {
    clientsById: { [key: string]: ClientType };
}

/** Create an immutable clients state
 * ClientType - generic type - client type being handled by this function
 */
export type ImmutableClientsState<ClientType> = ClientState<ClientType> &
    SeamlessImmutable.ImmutableObject<ClientState<ClientType>>;

/** initial clients-state state */
const initialState = SeamlessImmutable({
    clientsById: {},
});

/** factory function to create reducer
 * ClientType - generic type - client type being handled by this function
 */
export const reducerFactory = <ClientType>() =>
    /** the clients reducer function */
    function reducer(
        state: ImmutableClientsState<ClientType> = initialState,
        action: ClientsActionTypes<ClientType>,
    ): ImmutableClientsState<ClientType> {
        switch (action.type) {
            case CLIENTS_FETCHED:
                return SeamlessImmutable({
                    ...state,
                    clientsById: { ...state.clientsById, ...action.clientsById },
                });
            case REMOVE_CLIENTS:
                return SeamlessImmutable({
                    ...state,
                    clientsById: action.clientsById,
                });
            default:
                return state;
        }
    };

// Selectors

/** factory function that creates selector
 * ClientType - generic type - client type being handled by this function
 */
export const getClientsByIdFactory = <ClientType>(
    reducerName: string,
): ((state: Partial<Store>) => Dictionary<ClientType>) => {
    /** returns all clients in the store as values whose keys are their respective ids
     * @param {Partial<Store>} state - the redux store
     */
    return function(state: Partial<Store>): Dictionary<ClientType> {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        return (state as any)[reducerName].clientsById;
    };
};

/** factory function that creates selector
 * ClientType - generic type - client type being handled by this function
 */
export const getClientsArrayFactory = <ClientType>(reducerName: string) =>
    /** gets clients as an array of clients objects
     * @param {Partial<Store>} state - the redux store
     * @return {Client[]} - an array of clients objs
     */
    function(state: Partial<Store>): ClientType[] {
        const getClientsById = getClientsByIdFactory<ClientType>(reducerName);
        return values<ClientType>(getClientsById(state));
    };

/** factory function that creates selector
 * ClientType - generic type - client type being handled by this function
 */
export const getClientByIdFactory = <ClientType>(reducerName: string) =>
    /** get a specific client by their id
     * @param {Partial<Store>} state - the redux store
     * @return {Client | null} a client obj if the id is found else null
     */
    function(state: Partial<Store>, id: string): ClientType | null {
        return get(getClientsByIdFactory<ClientType>(reducerName)(state), id) || null;
    };
