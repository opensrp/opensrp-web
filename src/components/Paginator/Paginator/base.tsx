/** A bootstrap powered pagination component
 * inspired by https://scotch.io/tutorials/build-custom-pagination-with-react
 */
import React, { Fragment, MouseEvent } from 'react';
import { Pagination, PaginationItem, PaginationLink } from 'reactstrap';
import { fetchPageNumbers } from './utils';

/** interface for argument passed to the opPageChange prop */
export interface PaginationData {
  currentPage: number;
  pageLimit: number;
  totalPages: number;
  totalRecords: number;
}

/** interface for Paginator props */
export interface BasePaginatorProps {
  ariaLabel?: string;
  currentPage: number;
  endLabel: string;
  nextLabel: string;
  pageChangeHandler: (e: PaginationData) => void;
  pageLimit: number;
  pageNeighbours: number;
  previousLabel: string;
  startLabel: string;
  totalRecords: number;
}

/** default props for paginator component */
export const defaultBasePaginatorProps: BasePaginatorProps = {
  ariaLabel: 'page navigation',
  currentPage: 1,
  endLabel: 'End',
  nextLabel: 'Next',
  pageChangeHandler: f => f,
  pageLimit: 30,
  pageNeighbours: 2,
  previousLabel: 'Previous',
  startLabel: 'Start',
  totalRecords: 0,
};

/** the pagination component */
const BasePaginator = (props: BasePaginatorProps) => {
  const {
    totalRecords,
    pageLimit,
    pageNeighbours,
    ariaLabel,
    pageChangeHandler: onPageChange,
    currentPage,
  } = props;

  let totalPages = 0;

  // pageNeighbours can be in [0, 1, 2, 3, 4, 5]
  const neighbourPillsNum = Math.max(0, Math.min(pageNeighbours, 5));

  totalPages = Math.ceil(totalRecords / pageLimit);

  /** make sure passed in current page is within possible range */
  const sanitizedCurrentPage = Math.max(1, Math.min(currentPage, totalPages));

  const pages = fetchPageNumbers(neighbourPillsNum, totalPages, currentPage);

  // click handlers

  /** click Handler; registers the number of the clicked pagination Item
   * handles click event
   * @param {number} - number of the clicked paginationItem
   * @return {(e: MouseEvent) => void }
   */
  const handleClick = (page: number) => (e: MouseEvent) => {
    e.preventDefault();
    gotoPage(page);
  };

  /** click handler for when user clicks on previous
   * @param {MouseEvent} e - MouseEvent
   */
  const handleMoveLeft = (e: MouseEvent) => {
    e.preventDefault();
    gotoPage(currentPage - 1);
  };

  /** click handler for when user clicks on next
   * @param {MouseEvent} e - MouseEvent
   */
  const handleMoveRight = (e: MouseEvent) => {
    e.preventDefault();
    gotoPage(currentPage + 1);
  };

  /** called when user click on a certain pagination item
   * calls the onPageChange handler
   * @param {number} page -  the number of the page clicked
   */
  const gotoPage = (page: number) => {
    const thisPage = Math.max(1, Math.min(page, totalPages));
    const paginationData = {
      currentPage: thisPage,
      pageLimit,
      totalPages,
      totalRecords,
    };
    onPageChange(paginationData);
  };

  const { previousLabel, startLabel, nextLabel, endLabel } = props;

  return (
    <Fragment>
      <Pagination aria-label={ariaLabel} size="sm">
        <PaginationItem className={`page-item ${sanitizedCurrentPage > 1 ? '' : 'disabled'}`}>
          <PaginationLink
            className={`page-link`}
            href="#"
            aria-label={startLabel}
            onClick={handleClick(1)}
          >
            <span aria-hidden="true">{startLabel}</span>
            <span className="sr-only">{startLabel}</span>
          </PaginationLink>
        </PaginationItem>
        <PaginationItem className={`page-item ${sanitizedCurrentPage > 1 ? '' : 'disabled'}`}>
          <PaginationLink
            className={`page-link`}
            href="#"
            aria-label={previousLabel}
            onClick={handleMoveLeft}
          >
            <span aria-hidden="true">{previousLabel}</span>
            <span className="sr-only">{previousLabel}</span>
          </PaginationLink>
        </PaginationItem>
        {pages.map((page, index) => {
          return (
            <PaginationItem
              key={index}
              className={`page-item ${sanitizedCurrentPage === page ? ' active' : ''}`}
            >
              <PaginationLink className="page-link" href="#" onClick={handleClick(page as number)}>
                {page}
              </PaginationLink>
            </PaginationItem>
          );
        })}

        <PaginationItem
          className={`page-item  ${sanitizedCurrentPage < totalPages ? '' : 'disabled'}`}
        >
          <PaginationLink
            className={`page-link`}
            href="#"
            aria-label={nextLabel}
            onClick={handleMoveRight}
          >
            <span aria-hidden="true">{nextLabel}</span>
            <span className="sr-only">{nextLabel}</span>
          </PaginationLink>
        </PaginationItem>
        <PaginationItem
          className={`page-item ${sanitizedCurrentPage < totalPages ? '' : 'disabled'}`}
        >
          <PaginationLink
            className={`page-link ${sanitizedCurrentPage < totalPages ? '' : 'disabled'}`}
            href="#"
            aria-label={endLabel}
            onClick={handleClick(totalPages)}
          >
            <span aria-hidden="true">{endLabel}</span>
            <span className="sr-only">{endLabel}</span>
          </PaginationLink>
        </PaginationItem>
      </Pagination>
    </Fragment>
  );
};

BasePaginator.defaultProps = defaultBasePaginatorProps;

export { BasePaginator };
