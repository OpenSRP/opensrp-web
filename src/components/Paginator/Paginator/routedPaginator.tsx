/** Wraps basePagination with an interface that allows the basePagination
 * to work with the react-router.
 */
import React, { useEffect, useState } from 'react';
import { RouteComponentProps, withRouter } from 'react-router';
import { BasePaginator, PaginationData } from './base';

/** interface for RoutedPagination props */
export interface RoutedPaginatorProps {
  ariaLabel?: string;
  endLabel: string;
  nextLabel: string;
  pageLimit: number;
  pageNeighbours: number;
  previousLabel: string;
  startLabel: string;
  totalRecords: number;
  urlKey: string;
}

/** default props for routedPagination component */
const defaultRoutedPaginatorProps: RoutedPaginatorProps = {
  ariaLabel: 'page navigation',
  endLabel: 'End',
  nextLabel: 'Next',
  pageLimit: 30,
  pageNeighbours: 2,
  previousLabel: 'Previous',
  startLabel: 'Start',
  totalRecords: 0,
  urlKey: 'page',
};

/** The RoutedPaginator dumb component */
const Paginator = (
  props: RoutedPaginatorProps & RouteComponentProps<{ [key: string]: string }>
) => {
  const [currentPage, setCurrentPage] = useState<number>(1);
  const { urlKey } = props;

  useEffect(() => {
    // get currentPage from route and set that as the currentPage
    currentPageFromRoute();
  }, []);

  // functions

  /** gets the current page from url, this assumes the urls are formatted
   * in a certain way.
   */
  const currentPageFromRoute = () => {
    const { match } = props;
    const pageFromUrl = match.params[urlKey];
    const redirectToPage: number = pageFromUrl ? parseInt(pageFromUrl, 10) : 1;
    setCurrentPage(redirectToPage);
  };

  /** takes in a number and modifies the url such that redirecting to the new url
   * takes you to that page number
   */
  const getRedirectUrl = (
    thisPage: number,
    path: string,
    params: { [key: string]: string | undefined }
  ) => {
    // create new url to redirect to
    let urlToRedirectTo = '';
    Object.entries(params).forEach(([k, v]) => {
      let val: number | string | undefined = v;
      if (k === urlKey) {
        val = thisPage;
      }
      urlToRedirectTo = path.replace(`/:${k}`, `/${val}`);
    });
    return urlToRedirectTo;
  };

  /** changeHandler that gets called when user clicks on paginationItem
   *
   * @param {PaginationData} paginationData - state of pagination component
   */
  const pageChangeHandler = (paginationData: PaginationData) => {
    // get clicked route; modify the url and redirect to that.
    const { path, params } = props.match;

    const urlToRedirectTo = getRedirectUrl(paginationData.currentPage, path, params);
    // now redirect programmatically to new url
    props.history.push(urlToRedirectTo);
  };

  const {
    ariaLabel,
    endLabel,
    nextLabel,
    pageLimit,
    pageNeighbours,
    previousLabel,
    startLabel,
    totalRecords,
  } = props;

  /** const props for base paginator */
  const basePaginatorProps = {
    ariaLabel,
    currentPage,
    endLabel,
    nextLabel,
    pageChangeHandler,
    pageLimit,
    pageNeighbours,
    previousLabel,
    startLabel,
    totalRecords,
  };

  return <BasePaginator {...basePaginatorProps} />;
};

Paginator.defaultProps = defaultRoutedPaginatorProps;

const RoutedPaginator = withRouter(Paginator);

export { RoutedPaginator };
