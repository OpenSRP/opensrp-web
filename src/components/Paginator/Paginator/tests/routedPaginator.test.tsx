import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { createBrowserHistory } from 'history';
import React from 'react';
import { MemoryRouter, Route, Switch } from 'react-router';
import { Router } from 'react-router-dom';
import { BasePaginator } from '../base';
import { RoutedPaginator } from '../routedPaginator';

const history = createBrowserHistory();

const TestComponent = props => (
  <Switch>
    <Route
      exact={true}
      path="/records/:page"
      // tslint:disable-next-line: jsx-no-lambda
      render={routeProps => <RoutedPaginator {...routeProps} {...props} />}
    />
  </Switch>
);

describe('packages/paginator/routedPaginator', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('renders without crashing', () => {
    shallow(
      <Router history={history}>
        <RoutedPaginator />
      </Router>
    );
  });

  it('renders correctly', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/records/2']}>
        <TestComponent />
      </MemoryRouter>
    );

    const basePaginatorWrapper = wrapper.find(BasePaginator);
    const PIWrapper = basePaginatorWrapper.find('PaginationItem');
    const paginationItemNum = PIWrapper.length;
    expect(paginationItemNum).toEqual(4);
    expect(PIWrapper.at(0).hasClass('disabled')).toBeTruthy();
    expect(PIWrapper.at(1).hasClass('disabled')).toBeTruthy();
    expect(PIWrapper.at(paginationItemNum - 1).hasClass('disabled')).toBeTruthy();
    expect(PIWrapper.at(paginationItemNum - 2).hasClass('disabled')).toBeTruthy();
    expect(toJson(PIWrapper.at(0))).toMatchSnapshot('Start page');
    expect(toJson(PIWrapper.at(1))).toMatchSnapshot('previous page');
    expect(toJson(PIWrapper.at(paginationItemNum - 1))).toMatchSnapshot('Next page');
    expect(toJson(PIWrapper.at(paginationItemNum - 2))).toMatchSnapshot('End page');
  });

  it('passes corect page from url', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/records/2']}>
        <TestComponent />
      </MemoryRouter>
    );
    const basePaginatorWrapper = wrapper.find(BasePaginator);
    expect(basePaginatorWrapper.props().currentPage).toEqual(2);
  });

  it('renders correctly; edge case where page is gt than actual pages', () => {
    const props = { totalRecords: 5 };
    const wrapper = mount(
      <MemoryRouter initialEntries={['/records/2']}>
        <TestComponent {...props} />
      </MemoryRouter>
    );
    const basePaginatorWrapper = wrapper.find(BasePaginator);
    expect(basePaginatorWrapper.props().currentPage).toEqual(2);
    const PIWrapper = wrapper.find('PaginationItem');
    const paginationItemNum = PIWrapper.length;
    expect(paginationItemNum).toEqual(5);
    expect(PIWrapper.at(2).hasClass('active')).toBeTruthy();
  });

  it('clicking a page redirects to the correct page', () => {
    // 5 pages
    const props = { totalRecords: 135, pageLimit: 10 };
    const wrapper = mount(
      <MemoryRouter initialEntries={['/records/2']}>
        <TestComponent {...props} />
      </MemoryRouter>
    );
    const PIWrapper = wrapper.find('PaginationItem');
    // expect(toJson(wrapper.find(BasePaginator))).toMatchSnapshot('8 !== 7')
    expect(PIWrapper.length).toEqual(8);

    // go to the 1st page
    PIWrapper.at(2)
      .find('PaginationLink a')
      .simulate('click');

    wrapper.update();

    // inspect the pathname entry of the current location.
    const routedPaginatorWrapper = wrapper.find(RoutedPaginator);
    expect((routedPaginatorWrapper.props() as any).location.pathname).toEqual('/records/1');
  });
});
