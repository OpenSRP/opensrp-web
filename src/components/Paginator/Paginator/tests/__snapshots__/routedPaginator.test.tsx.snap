// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`packages/paginator/routedPaginator renders correctly: End page 1`] = `
<PaginationItem
  className="page-item  disabled"
  tag="li"
>
  <li
    className="page-item  disabled page-item"
  >
    <PaginationLink
      aria-label="Next"
      className="page-link"
      href="#"
      onClick={[Function]}
      tag="a"
    >
      <a
        aria-label="Next"
        className="page-link page-link"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          Next
        </span>
        <span
          className="sr-only"
        >
          Next
        </span>
      </a>
    </PaginationLink>
  </li>
</PaginationItem>
`;

exports[`packages/paginator/routedPaginator renders correctly: Next page 1`] = `
<PaginationItem
  className="page-item disabled"
  tag="li"
>
  <li
    className="page-item disabled page-item"
  >
    <PaginationLink
      aria-label="End"
      className="page-link disabled"
      href="#"
      onClick={[Function]}
      tag="a"
    >
      <a
        aria-label="End"
        className="page-link disabled page-link"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          End
        </span>
        <span
          className="sr-only"
        >
          End
        </span>
      </a>
    </PaginationLink>
  </li>
</PaginationItem>
`;

exports[`packages/paginator/routedPaginator renders correctly: Start page 1`] = `
<PaginationItem
  className="page-item disabled"
  tag="li"
>
  <li
    className="page-item disabled page-item"
  >
    <PaginationLink
      aria-label="Start"
      className="page-link"
      href="#"
      onClick={[Function]}
      tag="a"
    >
      <a
        aria-label="Start"
        className="page-link page-link"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          Start
        </span>
        <span
          className="sr-only"
        >
          Start
        </span>
      </a>
    </PaginationLink>
  </li>
</PaginationItem>
`;

exports[`packages/paginator/routedPaginator renders correctly: previous page 1`] = `
<PaginationItem
  className="page-item disabled"
  tag="li"
>
  <li
    className="page-item disabled page-item"
  >
    <PaginationLink
      aria-label="Previous"
      className="page-link"
      href="#"
      onClick={[Function]}
      tag="a"
    >
      <a
        aria-label="Previous"
        className="page-link page-link"
        href="#"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          Previous
        </span>
        <span
          className="sr-only"
        >
          Previous
        </span>
      </a>
    </PaginationLink>
  </li>
</PaginationItem>
`;
